plugins/cython/compile_cython.py

```diff
@@ -37,43 +44,66 @@ def product_types(cls):
     def cache_target_dirs(self):
         return True
 
-    @property
-    def create_target_dirs(self):
-        return True
-
-    def __init__(self, context, workdir):
-        super(CompileCythonCreate, self).__init__(context, workdir)
-
     def execute_codegen(self, target, results_dir):
-        self.context.log.info("Processing target %s" % target)
-
-        # Creating the pex can, as a side effect, change the location of the symbolic link pants creates
-        result = self.create_pex(PexInfo.default())
-        full_path = os.path.join(get_buildroot(), target.target_base)
-        # subprocess.check_call(['cmake', full_path], cwd=results_dir)
-        # subprocess.check_call(['make'], cwd=results_dir)
-
-        result_code = result.run(
-            with_chroot=True,
-            blocking=True,
-            args=(os.path.join(full_path, 'setup.py'), 'build_ext', '--inplace', '--verbose'),
-            # Passing PATH helps cython find the correct c++ compiler
-            env={'libraries': results_dir, 'PATH': os.getenv('PATH')}
-        )
-
-        if result_code != 0:
-            raise ValueError('creating cython library failed')
-
-        library_source_path = os.path.join(result.path(), target.output)
-
-        library_output = os.path.join(results_dir, target.output)
-        safe_mkdir_for(library_output)
-        shutil.move(library_source_path, library_output)
-
-        self.context.log.info('created library {}'.format(os.path.relpath(target.output, get_buildroot())))
-
-    def synthetic_target_type_by_target(self, target):
-        return Resources
+        self.context.log.info("Processing target {}".format(target))
+
+        requirements_pex = self.context.products.get_data(ResolveRequirements.REQUIREMENTS_PEX)
+
+        interpreter = self.context.products.get_data(PythonInterpreter)
+        pex_info = PexInfo.default(interpreter)
+        pex_info.pex_path = requirements_pex.path()
+        with temporary_dir() as source_pex_chroot:
+            sources_pex_builder = PEXBuilder(
+                path=source_pex_chroot,
+                interpreter=interpreter,
+                copy=True,
+                pex_info=pex_info
+            )
+            pex_build_util.dump_sources(sources_pex_builder, target, self.context.log)
+            sources_pex_builder.freeze()
+            codegen_pex = PEX(sources_pex_builder.path(), interpreter)
+
+            setup_py_paths = []
```

Comment:
Your code assumed `setup.py` was at a fixed location relative to the target. These gymnastics allow for, say:
```
app/
  native/
    BUILD
    setup.py
```
or:
```
app/
  BUILD
  native/
    setup.py
```
Where `app` is configured as a source root:
```ini
[source]
source_roots: {
    'app': 'python'
  }
```